larpandoracontent/LArTrackShowerId/TrackShowerIdFeatureTool.cc

```diff
@@ -326,19 +373,781 @@ StatusCode MipEnergyFeatureTool::ReadSettings(const TiXmlHandle xmlHandle)
+        try
+        {
+            ClusterList clusterList;
+            LArPfoHelper::GetThreeDClusterList(pInputPfo, clusterList);
+            if (1 == clusterList.size())
+            {
+                const Cluster *const pCluster(clusterList.front());
+                const ThreeDSlidingFitResult sliding3DFitResult(pCluster, m_slidingLinearFitWindow, LArGeometryHelper::GetWireZPitch(this->GetPandora()));
+                const TwoDSlidingFitResult slidingFitResult(sliding3DFitResult.GetFirstFitResult());
+                float straightLineLength = (slidingFitResult.GetGlobalMaxLayerPosition() - slidingFitResult.GetGlobalMinLayerPosition()).GetMagnitude();
+                if (straightLineLength > std::numeric_limits<double>::epsilon())
+                    ratio = vertexDistance / straightLineLength;
+            }
+        }
+        catch (const StatusCodeException &)
+        {
+            ratio = -1.f;
+        }
+        featureVector.push_back(ratio);
+    }
+    else
+    {
+        featureVector.push_back(vertexDistance);
+    }
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+StatusCode ThreeDVertexDistanceFeatureTool::ReadSettings(const TiXmlHandle xmlHandle)
+{
+    PANDORA_RETURN_RESULT_IF_AND_IF(STATUS_CODE_SUCCESS, STATUS_CODE_NOT_FOUND, !=, XmlHelper::ReadValue(xmlHandle,
+        "RatioVariables", m_ratioVariables));
+
+    PANDORA_RETURN_RESULT_IF_AND_IF(STATUS_CODE_SUCCESS, STATUS_CODE_NOT_FOUND, !=, XmlHelper::ReadValue(xmlHandle,
+        "SlidingLinearFitWindow", m_slidingLinearFitWindow));
+
+    return STATUS_CODE_SUCCESS;
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+OpeningAngleFeatureTool::OpeningAngleFeatureTool() 
+{
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+void OpeningAngleFeatureTool::Run(SupportVectorMachine::DoubleVector &featureVector, const Algorithm *const pAlgorithm,
+    const pandora::ParticleFlowObject *const pInputPfo)
+{
+    if (PandoraContentApi::GetSettings(*pAlgorithm)->ShouldDisplayAlgorithmInfo())
+        std::cout << "----> Running Algorithm Tool: " << this->GetInstanceName() << ", " << this->GetType() << std::endl;
+
+    // Need the 3D clusters and hits to calculate PCA components
+    ClusterList threeDClusterList;
+    LArPfoHelper::GetThreeDClusterList(pInputPfo, threeDClusterList);
+
+    CaloHitList threeDCaloHitList; // Might be able to get through LArPfoHelper instead
+    const Cluster *const pThreeDCluster(threeDClusterList.front());
+    pThreeDCluster->GetOrderedCaloHitList().FillCaloHitList(threeDCaloHitList);
```

Comment:
Should now be able to get all 3D hits directly from the Pfo, using the LArPfoHelper.

Reply:
ok, I'll have a look.